pyro/contrib/funsor/named_messenger.py

```diff
@@ -0,0 +1,154 @@
+
+        # every markov context is responsible for knowing about non-fresh local vars,
+        # and fresh vars are handled by passing information up & down the stack
+
+        # handling non-fresh vars: just look at the deepest context they appear in
+        name_to_dim.update({name: self.current_frame.name_to_dim[name]
+                            for name in funsor_value.inputs
+                            if name in self.current_frame.name_to_dim
+                            and name not in name_to_dim})
+
+        # handling fresh vars: pass responsibility for allocation up the stack
+        depth = msg.setdefault("depth", -self.history - 1)
+        if depth == -self.history - 1:
+            msg["fresh"] = frozenset(name for name in funsor_value.inputs
+                                     if name not in name_to_dim
+                                     and name not in self.current_frame.name_to_dim)
+        fresh = msg["fresh"]
+
+        for pos in reversed(range(max(0, self._pos - self.history), self._pos + 1)):
+            frame = self._stack[pos]
+            msg["depth"] += 1
+            if msg["depth"] == 0:
+                for name in fresh:
+                    if name not in frame.name_to_dim:
+                        dim = -1
+                        while dim in frame.dim_to_name:
+                            dim -= 1
+                        frame.name_to_dim[name] = dim
+                    dim = frame.name_to_dim[name]
+                    frame.dim_to_name[dim] = (name, funsor_value.inputs[name])
+                    name_to_dim[name] = dim
+        msg['depth'] += self._pos
+
+    def _pyro_post_to_data(self, msg):
+        funsor_value, name_to_dim = msg["args"]
+
+        msg['depth'] -= self._pos
+
+        # copy fresh variables down the stack
+        for pos in range(max(0, self._pos - self.history), self._pos + 1):
+            frame = self._stack[pos]
+            if msg["depth"] <= 0 and msg["depth"] >= -self.history - 1:
+                frame.name_to_dim.update({name: name_to_dim[name] for name in msg["fresh"]})
+                # dim_to_name maps dims to (name, domain) pairs
+                frame.dim_to_name.update(
+                    {name_to_dim[name]: (name, funsor_value.inputs[name]) for name in msg["fresh"]})
+
+            msg["depth"] -= 1
+
+    def _pyro_to_funsor(self, msg):
+        if len(msg["args"]) < 3:
+            msg["args"] = msg["args"] + (OrderedDict(),)
+        raw_value, output, dim_to_name = msg["args"]
+
+        event_dim = len(output.shape)
+        batch_dim = len(raw_value.shape) - event_dim
+
+        # since we don't allow fresh positional dims, only look at current frame
+        dim_to_name.update({dim: self.current_frame.dim_to_name[dim]
+                            for dim in range(-batch_dim, 0)
+                            if dim not in dim_to_name
+                            and dim in self.current_frame.dim_to_name})
```

Comment:
This method is the simplest place to start understanding what the current implementation of `NamedMessenger` is doing: inference code calls `pyro.contrib.funsor.to_funsor` on a `torch.Tensor` or `TorchDistribution` with no `dim_to_name` argument, and the most deeply nested `NamedMessenger` intercepts that call and populates a `dim_to_name` mapping with information about the meaning of batch dimensions collected from all previous `pyro.contrib.funsor.to_data` calls it or its parents have seen.